user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts

```diff
@@ -0,0 +1,159 @@
+import {ResponseError} from '@nu-art/ts-common/core/exceptions/types';
+import {dispatch_OnLoginFailed} from '../../login-attempts/backend/dispatchers';
+import {OnUserLogin} from '../../account/backend';
+import {SafeDB_Account} from '../../account/shared';
+
+
+type Config = DBApiConfigV3<DBProto_FailedLoginAttempt> & {
+	loginBlockedTTL: number;
+	maxLoginAttempts: number;
+	documentTTL: number;
+}
+
+type LoginBlockedErrorBody = ResponseError<typeof ErrorType_LoginBlocked, {
+	blockedUntil: number;
+}>
+
+/**
+ * Manage and handle login failed attempts,
+ * Default login blocked timer is 5 minutes
+ */
+export class ModuleBE_FailedLoginAttemptDB_Class
+	extends ModuleBE_BaseDB<DBProto_FailedLoginAttempt, Config> implements OnUserLogin {
+
+	constructor() {
+		super(DBDef_FailedLoginAttempt);
+		this.setDefaultConfig({
+			loginBlockedTTL: 5 * Minute, // default login block is 5 minutes
+			maxLoginAttempts: DefaultMaxLoginAttempts,
+			documentTTL: 3 * Minute
+		});
+	}
+
+	__onUserLogin(account: SafeDB_Account) {
+		return this.isAccountLoginBlocked(account._id);
+	}
+
+	//######################### Public Logic #########################
+
+	/**
+	 * Failed login attempt handler, will take care of updating or managing
+	 * failed login attempt event.
+	 * @param accountId The account that failed login.
+	 */
+	public updateFailedLoginAttempt = async (accountId: UniqueId) => {
+		// Fetch the existing attempt
+		const existingLoginAttempt = await this.getExistingLoginAttempt(accountId);
+
+		if (!existingLoginAttempt) {
+			return this.createNewLoginAttempt(accountId);
+		}
+
+		// Check if the login is already blocked
+		if (this.isLoginBlocked(existingLoginAttempt)) {
+			return this.handleBlockedLogin(existingLoginAttempt);
+		}
+
+		// Check if TTL has expired
+		if (this.isTTLExpired(existingLoginAttempt) || exists(existingLoginAttempt.loginSuccessfulAt)) {
+			return this.createNewLoginAttempt(accountId);
+		}
+
+		// Update the failed login attempt
+		return this.incrementLoginAttempt(existingLoginAttempt);
+	};
+
+	//######################### Internal Logic #########################
+
+	/**
+	 * After a successful login attempt use this function to validate if user isn't blocked
+	 * or if block time elapsed
+	 * @param accountId The account that successfully logged in
+	 */
+	private isAccountLoginBlocked = async (accountId: UniqueId) => {
+		const loginAttempt = await this.getExistingLoginAttempt(accountId);
+
+		// fail fast if there's no login attempts document
+		if (!loginAttempt)
+			return;
+
+		// update login timestamp if login is successful and not blocked
+		if (loginAttempt.count < this.config.maxLoginAttempts) {
+			loginAttempt.loginSuccessfulAt = currentTimeMillis();
+			return this.set.item(loginAttempt);
+		}
+
+		// if the login is still blocked throw blocked error
+		if (this.assertLoginBlock(loginAttempt))
+			return this.handleBlockedLogin(loginAttempt);
+	};
```

Comment:
not sure how this method makes sense..
this suppose to be a onLoginSuccessful, no?

Reply:
I call it from the on login successfull dispatcher.. it's just this module implementation.. I just exctrated it to a function in the module that's if...